[docs] Describe type-casting HTML elements in client side TypeScript code (#6047)
DevExpress/testcafe: docs/articles/documentation/guides/concepts/typescript-and-coffeescript.md

> and [selector.addCustomMethods](../../reference/test-api/selector/addcustommethods.md)
> sections to learn how to extend selectors in TypeScript.

#### Type Cast Page Elements

> When you grab HTML elements in your TypeScript [client-side code](../basic-guides/obtain-client-side-info.md), you need to cast the elements to the [HTMLElement](https://www.typescriptlang.org/docs/handbook/dom-manipulation.html#an-exploration-into-the-htmlelement-type) type. To avoid type errors, cast the element to [unknown](https://www.typescriptlang.org/docs/handbook/2/functions.html#unknown) and then to [HTMLElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement).

Reviewer: What about such variant?

```js
When you grab HTML elements in your TypeScript client-side code, you need to cast the element to an appropriate type.  Choose the type depending on your needs. If in code your need only basic HTMLElement element method and properties when cast to HTMLElement type.
See example below
<example>

To use methods and properties of a certain HTMLElement interface (for example, HTMLOListElement), cast the element to that specific type.
<example>

```

> An example below uses [Element.scrollIntoView()](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) inside a [client function](../../reference/test-api/clientfunction/README.md) to scroll an HTML element into view.

Author: ```suggestion
A [client function](../../reference/test-api/clientfunction/README.md) in the example below calls the [Element.scrollIntoView()](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) method to scroll an element into view.
```

```ts
import { ClientFunction, Selector } from 'testcafe';

const scrollIntoView = ClientFunction( (selector: Selector) => {
    const element = selector() as unknown as HTMLElement;
    element.scrollIntoView();
});

fixture`HTMLElement`
    .page('https://example.com');

test('Scroll element into view', async t => {
    const bottomOfPage = Selector('#bottom-div');

    await scrollIntoView(bottomOfPage);
});
```

To use members of a particular HTML element interface (for example, [HTMLOListElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement)), cast the element to that specific type.

> The example below uses [t.eval](../../reference/test-api/testcontroller/eval.md) to determine if an ordered list has the [*reversed* property](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement#properties). Because the property exists on the [HTMLOListElement Interface](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement#properties), but not on the [HTMLElement Interface](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement#properties), a narrower type casting is needed.

Reviewer 2: Very confusing. Also, only the type itself can be narrow - not the casting.

```suggestion
The example below calls the [t.eval](../../reference/test-api/testcontroller/eval.md) method to determine if an ordered list is [reversed](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement#properties). The generic [HTMLElement Interface](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement#properties) cannot access the DOM element's `reversed` property, so the `list` object is assigned the [HTMLOListElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement#properties) data type.
```

Author: > The generic HTMLElement Interface cannot access the DOM element's `reversed` property

An interface can't access a property. A user might access a property defined on an interface.

```suggestion
The code in an example below calls the [t.eval](../../reference/test-api/testcontroller/eval.md) method to determine if an ordered list is [reversed](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement#properties). The generic [HTMLElement Interface](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement#properties) does not have a `reversed` property, so it is necessary to cast the [HTMLOListElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement#properties) data type to the `list` object to avoid a compilation error.
```

Reviewer 2: Almost! Just a couple notes. 

>An interface can't access a property.
agreed, although it can provide access to a property.
> A user might access a property defined on an interface.
not necessarily - the property belongs to an object, the interface is merely a mediator. Lists have the `reversed` property whether or not we can access it.

>cast the type to the list
It's actually the opposite. You cast the list to be of a certain type. see [this](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/types/casting-and-type-conversions) example. If possible, I'd completely avoid using the word "cast" since it's *not* the official term for this process. The TypeScript handbook uses the verb "to convert" interchangeably 

```suggestion
The example code below calls the [t.eval](../../reference/test-api/testcontroller/eval.md) method to determine if an ordered list is [reversed](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement#properties). The generic [HTMLElement Interface](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement#properties) does not provide access to the element's `reversed` property. Therefore it is necessary to convert the `list` object to the [HTMLOListElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement#properties) data type.
```

Author: > It's actually the opposite

You are right, that's a typo

Just one thing: 

```suggestion
The example code below calls the [t.eval](../../reference/test-api/testcontroller/eval.md) method to determine if an ordered list is [reversed](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement#properties). The generic [HTMLElement Interface](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement#properties) does not provide access to the element's `reversed` property. To avoid compilation errors, it is necessary to convert the `list` object to the [HTMLOListElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement#properties) data type.
```

> and [selector.addCustomMethods](../../reference/test-api/selector/addcustommethods.md)
> sections to learn how to extend selectors in TypeScript.

#### Type Cast Page Elements

> TypeScript compilers can not automatically identify TestCafe objects that refer to DOM elements. Perform manual [type assertions](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions) in your [client-side code](../basic-guides/obtain-client-side-info.md) to ensure correct TypeScript compilation.

Reviewer 3: ```suggestion
TypeScript compilers cannot automatically identify TestCafe objects that refer to DOM elements. Perform manual [type assertions](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions) in your [client-side code](../basic-guides/obtain-client-side-info.md) to ensure correct TypeScript compilation.
```

Specify the [HTMLElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement) data type to access the DOM element's generic HTMLElement interface.

A [client function](../../reference/test-api/clientfunction/README.md) in the example below calls the [Element.scrollIntoView()](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) method to scroll an element into view.

```ts
import { ClientFunction, Selector } from 'testcafe';

const scrollIntoView = ClientFunction( (selector: Selector) => {
    const element = selector() as unknown as HTMLElement;
    element.scrollIntoView();
});

fixture`HTMLElement`
    .page('https://example.com');

test('Scroll element into view', async t => {
    const bottomOfPage = Selector('#bottom-div');

    await scrollIntoView(bottomOfPage);
});
```

> Pick element-specific data types, such as [HTMLOListElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement)), whenever appropriate.

Reviewer 3: whenever appropriate 
is that part necessary

Author: Doesn't really explain much. Does this work:

```suggestion
To avoid compilation errors, pick element-specific data types, such as [HTMLOListElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOListElement)).
```
